sdk/elixir/lib/dagger/internal/engine/downloader.ex

```diff
@@ -26,21 +26,87 @@ defmodule Dagger.Internal.Engine.Downloader do
     end
   end
 
-  # TODO: checksum.
   defp extract_cli(cli_host, cli_version, bin_path) do
-    with {:ok, response} <- Req.get(cli_archive_url(cli_host, cli_version)) do
+    req = Req.new(url: cli_archive_url(cli_host, cli_version))
+
+    with {:ok, raw_response} <- Req.get(req, raw: true),
+         :ok <- verify_checksum(raw_response, cli_host, cli_version) do
+      {_, decoded_response} = Req.Steps.decode_body({req, raw_response})
+
       {_, dagger_bin} =
-        Enum.find(response.body, fn {filename, _bin} ->
+        Enum.find(decoded_response.body, fn {filename, _bin} ->
           filename == dagger_bin_in_archive(os())
         end)
 
       File.write(bin_path, dagger_bin)
     end
   end
 
+  defp verify_checksum(raw_response, cli_host, cli_version) do
+    calculated_checksum = :crypto.hash(:sha256, raw_response.body) |> Base.encode16(case: :lower)
+
+    case expected_checksum(cli_host, cli_version) do
+      {:ok, expected_checksum} ->
+        if(calculated_checksum != expected_checksum) do
+          {:error, "checksum mismatch: expected #{expected_checksum}, got #{calculated_checksum}"}
+        else
+          :ok
+        end
+
+      error ->
+        error
+    end
+  end
+
+  defp expected_checksum(cli_host, cli_version) do
+    archive_name = default_cli_archive_name(os(), arch(), cli_version)
+
+    {:ok, checksum_map} = checksum_map(cli_host, cli_version)
+
+    expected_value =
+      Enum.find_value(checksum_map, fn {key, value} ->
+        if to_string(key) == archive_name, do: value
+      end)
+
+    if expected_value == nil do
+      {:error, "expected value find error"}
+    else
+      {:ok, expected_value}
+    end
+  end
+
+  defp checksum_map(cli_host, cli_version) do
+    try do
+      with {:ok, response} <- Req.get(checksum_url(cli_host, cli_version)) do
+        checksum_map =
+          response.body
+          |> String.split("\n", trim: true)
+          |> Stream.map(&String.split/1)
+          |> Enum.map(fn list ->
```

Comment:
This should be `Stream`?

Reply:
Overlooked. I used stream enumeration, thinking that the first fetch would be a large binary set. I later replaced it with enum but this line was overlooked. Stream is lazy unless there are large sets of binaries.

fixed.